### ISSUE TITLE
Lifecycle management of decorators

### ISSUE THREAD

**Issue by Dreamsorcerer** [aio-libs/aiocache#609]
Cache objects should be closed at the end of an application lifecycle, with `await Cache.close()`, or using `async with`.

The current design of decorators creates a new cache instance with each decorator and no attempt to close it is made, thus failing to manage this lifecycle management properly.

e.g.
```
@cached(...)  # New cache is created here, but never closed.
def foo(): ...
```

We also want to consider using aiojobs for managing some background tasks, but this additionally requires being created within a running loop, something which is unlikely when a decorator is called.

----

One solution I can think of, is to explicitly manage the caches, and pass them to the decorators. This may also need a `.start()` method to initiate the cache later. e.g.

```
cache = Cache(...)

@cached(cache)
def foo(): ...

async def main():
    # Initialise application
    cache.start()
    # Run application
    ...
    # Cleanup application
    await cache.close()
```

Or more succinctly:
```
async def main():
    async with cache:
        # Run application
```

**Comment by Dreamsorcerer** [comment]
These are causing warnings in our examples as well: #610.

**Comment by cancan101** [comment]
>The current design of decorators creates a new cache instance with each decorator 

Is that true if an alias is used? ie if I use the same alias for multiple decorators will the underlying `Cache` instance be shared?

**Comment by Dreamsorcerer** [comment]
> > The current design of decorators creates a new cache instance with each decorator
> 
> Is that true if an alias is used? ie if I use the same alias for multiple decorators will the underlying `Cache` instance be shared?

Yes, that should be shared, but the issue is really around the lifecycle of the cache, not the sharing. With an alias, the cache is still created on-demand and users are unlikely to think about closing the cache at any point.

I also don't see any real advantage to an alias like that, compared with just passing an explicit reference to the cache object directly.

**Comment by mavwolverine** [comment]
Won't closing cache on application exit be a problem in an autoscaled environment?

**Comment by elecnix** [comment]
Being able to close all caches would be nice.

With FastAPI:
    
    @app.on_event("shutdown")
    async def close_all_caches():
        for cache in aiocache.caches:
            try:
                await cache.close()
            except Exception:
                pass

**Comment by Dreamsorcerer** [comment]
I think that's trivial for the user to track the caches and do that themselves anyway. Doing it at the library level would likely cause issues if other libraries used aiocache and didn't expect to have their cache closed by another library or application.

**Comment by sakosha** [comment]
Hello @Dreamsorcerer!

Is this the only blocker for other work for 1.0 release?
Did you come up with vision of what should the decorator API look like?
